File: skript.js
const fs = require('fs').promises;
const path = require('path');
const xmlbuilder = require('xmlbuilder');

async function processJsonFile(filePath, output, test, maxItems) {
    try {
        const data = await fs.readFile(filePath, 'utf-8');
        const jsonData = JSON.parse(data);

        if (typeof jsonData !== 'object' || !jsonData.vehicle || !jsonData.categories) {
            console.log(`Skipping file ${filePath}: not in expected format`);
            return output;
        }

        const vehicleName = jsonData.vehicle.name;
        for (const category of jsonData.categories) {
            output = await extractSpareParts(category, [vehicleName, category.name], output, test, maxItems);
        }
    } catch (err) {
        console.error(`Error processing file ${filePath}: ${err}`);
    }
    return output;
}

async function extractSpareParts(data, parentNames, output, test, maxItems) {
    if (test && Object.keys(output).length >= maxItems) {
        return output;
    }

    if (data.spare_parts) {
        for (const part of data.spare_parts) {
            if (test && Object.keys(output).length >= maxItems) {
                return output;
            }

            const partName = part.product.name;
            const partNo = part.product.product_no;
            const vatPercent = part.product.vat_percent || 0;
            const unitPriceInclVat = part.product.unit_price_incl_vat || 0;

            if (parentNames.includes(null) || !partName || !partNo) {
                continue;
            }

            const categoryPath = parentNames.join(' > ');
            if (output[partNo]) {
                output[partNo].categories.push(categoryPath);
            } else {
                output[partNo] = {
                    name: partName,
                    categories: [categoryPath],
                    vat_percent: vatPercent,
                    unit_price_incl_vat: unitPriceInclVat
                };
            }
        }
    }

    if (data.categories) {
        for (const category of data.categories) {
            output = await extractSpareParts(category, parentNames.concat([category.name]), output, test, maxItems);
        }
    }

    return output;
}

function createXml(output, xmlFile) {
    const shop = xmlbuilder.create('SHOP');

    for (const partNo in output) {
        const info = output[partNo];
        const shopItem = shop.ele('SHOPITEM');
        shopItem.ele('NAME', info.name);
        shopItem.ele('CODE', partNo);

        const categoriesElem = shopItem.ele('CATEGORIES');
        info.categories.forEach(categoryPath => {
            const parts = categoryPath.split(' > ', 2);
            if (parts.length > 1) {
                const mainCategory = parts[0].replace(' / ', ' > ');
                const subCategory = parts[1];
                categoriesElem.ele('CATEGORY', `${mainCategory} > ${subCategory}`);
            } else {
                categoriesElem.ele('CATEGORY', parts[0]);
            }
        });

        shopItem.ele('PRICE_VAT', info.unit_price_incl_vat);
        shopItem.ele('VAT', info.vat_percent);
    }

    const xmlString = shop.end({ pretty: true, indent: '  ', newline: '\n' });
    return fs.writeFile(xmlFile, xmlString, 'utf-8');
}

async function main(folderPath, xmlFile, test = false, maxItems = 10) {
    let output = {};
    try {
        const files = await fs.readdir(folderPath);

        for (const file of files) {
            if (file.endsWith('.json')) {
                const filePath = path.join(folderPath, file);
                output = await processJsonFile(filePath, output, test, maxItems);
                if (test && Object.keys(output).length >= maxItems) {
                    break;
                }
            }
        }

        await createXml(output, xmlFile);
        console.log(`XML file ${xmlFile} has been created successfully.`);
    } catch (err) {
        console.error(`Error in main function: ${err}`);
    }
}

const folderPath = 'spare_parts_feed';  // Path to the folder with data
const xmlFile = 'output.xml';  // Name of the output XML file
const test = false;  // True for testing XML validation (generates only 10 items)
main(folderPath, xmlFile, test);

File: skript.py
import json
import os
import xml.etree.ElementTree as ET
from xml.dom import minidom

def process_json_file(file_path, output, test, max_items):
    with open(file_path, 'r', encoding='utf-8') as file:
        data = json.load(file)
        if not isinstance(data, dict) or "vehicle" not in data or "categories" not in data:
            print(f"Skipping file {file_path}: not in expected format")
            return
        vehicle_name = data["vehicle"]["name"]
        for category in data["categories"]:
            extract_spare_parts(category, [vehicle_name, category["name"]], output, test, max_items)

def extract_spare_parts(data, parent_names, output, test, max_items):
    if test and len(output) >= max_items:
        return

    if "spare_parts" in data:
        for part in data["spare_parts"]:
            if test and len(output) >= max_items:
                return
            part_name = part["product"]["name"]
            part_no = part["product"]["product_no"]
            vat_percent = part["product"].get("vat_percent", 0)
            unit_price_incl_vat = part["product"].get("unit_price_incl_vat", 0) or 0
            if None in parent_names or part_name is None or part_no is None:
                continue
            category_path = " > ".join(parent_names)
            if part_no in output:
                output[part_no]["categories"].append(category_path)
            else:
                output[part_no] = {
                    "name": part_name,
                    "categories": [category_path],
                    "vat_percent": vat_percent,
                    "unit_price_incl_vat": unit_price_incl_vat
                }
    
    if "categories" in data:
        for category in data["categories"]:
            extract_spare_parts(category, parent_names + [category["name"]], output, test, max_items)

def create_xml(output, xml_file):
    shop = ET.Element("SHOP")
    for part_no, info in output.items():
        shop_item = ET.SubElement(shop, "SHOPITEM")
        name = ET.SubElement(shop_item, "NAME")
        name.text = info["name"]
        code = ET.SubElement(shop_item, "CODE")
        code.text = part_no
        categories_elem = ET.SubElement(shop_item, "CATEGORIES")
        for category_path in info["categories"]:
            parts = category_path.split(" > ", 1)
            if len(parts) > 1:
                main_category = parts[0].replace(" / ", " > ")
                sub_category = parts[1]
                category = ET.SubElement(categories_elem, "CATEGORY")
                category.text = f"{main_category} > {sub_category}"
            else:
                category = ET.SubElement(categories_elem, "CATEGORY")
                category.text = parts[0]
        
        price_vat = ET.SubElement(shop_item, "PRICE_VAT")
        price_vat.text = str(info["unit_price_incl_vat"])
        vat = ET.SubElement(shop_item, "VAT")
        vat.text = str(info["vat_percent"])

    pretty_xml = prettify_xml(shop)
    with open(xml_file, 'w', encoding='utf-8') as file:
        file.write(pretty_xml)

def prettify_xml(elem):
    rough_string = ET.tostring(elem, 'utf-8')
    reparsed = minidom.parseString(rough_string)
    return reparsed.toprettyxml(indent="  ")

def main(folder_path, xml_file, test=False, max_items=10):
    output = {}
    for filename in os.listdir(folder_path):
        if filename.endswith(".json"):
            file_path = os.path.join(folder_path, filename)
            process_json_file(file_path, output, test, max_items)
            if test and len(output) >= max_items:
                break

    create_xml(output, xml_file)

if __name__ == "__main__":
    folder_path = 'spare_parts_feed'  # Path to the folder with data
    xml_file = 'output.xml'  # Name of the output XML file
    test = False  # True for testing XML validation (generates only 10 items)
    main(folder_path, xml_file, test)
